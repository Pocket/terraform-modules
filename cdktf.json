{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
