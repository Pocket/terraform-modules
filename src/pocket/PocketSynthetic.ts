import { Resource } from 'cdktf';
import { Synthetics, IAM, CloudWatch } from '@cdktf/provider-aws';
import { Construct } from 'constructs';
import { DataArchiveFile } from '@cdktf/provider-archive';

import compile from 'string-template/compile';
import fs = require('fs');
import path = require('path');

export interface PocketSyntheticProps {
  config?: {
    url: string;
    path: string;
    port: string;
    protocol: string;
  };
  check?: any;
  environment: string;
  artifactS3Location: string;
  pagerDutyArn?: string;
}

const defaultMonitor = compile(`
var synthetics = require('Synthetics');

const apiCanary = async function () {

    // Handle validation for positive scenario
    const validateSuccessful = async function (res) {
        return new Promise((resolve, reject) => {
            if (res.statusCode < 200 || res.statusCode > 299) {
                throw res.statusCode + ' ' + res.statusMessage;
            }

            let responseBody = '';
            res.on('data', (d) => {
                responseBody += d;
            });

            res.on('end', () => {
                // Add validation on 'responseBody' here if required.
                resolve();
            });
        });
    };

    let commonProps = {
        hostname: '{0}',
        method: 'GET',
        protocol: '{1}',
        port: '{2}',
        region: process.env.AWS_REGION,
        body:  "",
        headers: {
            "User-Agent": synthetics.getCanaryUserAgentString(),
        },
    }

    let allConfig = {
        includeRequestHeaders: true,
        includeResponseHeaders: true,
        includeRequestBody: true,
        includeResponseBody: false,
        restrictedHeaders: ["authorization", "x-amz-security-token"],
        continueOnHttpStepFailure: true
    };

    await synthetics.executeHttpStep('Verify synthetic check', {
        ...commonProps,
        path: '{3}',
    }, validateSuccessful, allConfig);
};

exports.handler = async () => {
    return await apiCanary();
};
`);

export class PocketSyntheticCheck extends Resource {
  constructor(
    scope: Construct,
    private name: string,
    private config: PocketSyntheticProps
  ) {
    super(scope, name);

    if (this.config.check == undefined) {
      this.config.check = defaultMonitor(
        this.config.config.url,
        this.config.config.protocol,
        this.config.config.port,
        this.config.config.path
      );
    }

    const indexPath = path.join(
      __dirname,
      'synthetics',
      'nodejs',
      'node_modules',
      'index.js'
    );

    fs.writeFileSync(indexPath, this.config.check);

    const zipFile = new DataArchiveFile(this, 'synthetic-zip-file', {
      type: 'zip',
      // TODO: hmm, maybe don't want to cause a diff evertyime
      outputPath: `index-${(+new Date()).toString(36)}.zip`,
      sourceDir: path.join(__dirname, 'synthetics'),
    });

    new Synthetics.SyntheticsCanary(this, `canary-${this.name}`, {
      name: `${this.name}-syntheic`.toLowerCase(),
      schedule: { expression: 'rate(3 minutes)' },
      runtimeVersion: 'syn-nodejs-puppeteer-3.3',
      artifactS3Location: this.config.artifactS3Location,
      zipFile: zipFile.outputPath,
      handler: 'index.handler',
      executionRoleArn: this.syntheticsExecutionRole(
        this.config.artifactS3Location
      ).arn,
      startCanary: true,
    });

    new CloudWatch.CloudwatchMetricAlarm(this, `synthetics-failed-alarm`, {
      alarmName: `${this.name}-SyntheticsFailed-Alarm`,
      namespace: 'CloudWatchSynthetics',
      metricName: 'Failures',
      dimensions: {
        CanaryName: `${this.name}-synthetic`.toLowerCase(),
      },
      period: 60,
      evaluationPeriods: 1,
      statistic: 'SampleCount',
      comparisonOperator: 'GreaterThanOrEqualToThreshold',
      threshold: 1,
      alarmDescription: 'Synthetic Check is failing',
      insufficientDataActions: [],
      alarmActions: [], //TODO: put a pagerduty arn here
      treatMissingData: 'notBreaching',
    });
  }

  private syntheticsExecutionRole(bucket: string) {
    const iamRole = new IAM.IamRole(this, 'synthetics-execution-role', {
      name: `synthetics-role-for-${this.name}`,
      assumeRolePolicy: this.syntheticsAssumePolicyDocument(),
    });

    const policy = new IAM.IamPolicy(this, 'execution-policy', {
      name: `${this.name}-ExecutionRolePolicy`,
      policy: this.getPolicyDocument(bucket.replace('s3://', '')),
    });

    new IAM.IamRolePolicyAttachment(this, 'execution-role-policy-attachment', {
      role: iamRole.name,
      policyArn: policy.arn,
      dependsOn: [iamRole, policy],
    });
    return iamRole;
  }

  private getPolicyDocument(bucket: string) {
    const document = {
      version: '2012-10-17',
      statement: [
        {
          effect: 'Allow',
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents',
          ],
          resources: ['arn:aws:logs:*:*:*'],
        },
        {
          effect: 'Allow',
          actions: [
            's3:ListAllMyBuckets',
            'cloudwatch:PutMetricData',
            'xray:PutTraceSegments',
          ],
          resources: ['*'],
        },
        {
          effect: 'Allow',
          actions: ['s3:PutObject', 's3:GetBucketLocation'],
          resources: [`arn:aws:s3:::${bucket}`, `arn:aws:s3:::${bucket}/*`],
        },
      ],
    };

    return new IAM.DataAwsIamPolicyDocument(
      this,
      'execution-policy-document',
      document
    ).json;
  }

  private syntheticsAssumePolicyDocument() {
    return new IAM.DataAwsIamPolicyDocument(this, 'assume-policy-document', {
      version: '2012-10-17',
      statement: [
        {
          effect: 'Allow',
          actions: ['sts:AssumeRole'],
          principals: [
            {
              identifiers: ['lambda.amazonaws.com', 'synthetics.amazonaws.com'],
              type: 'Service',
            },
          ],
        },
      ],
    }).json;
  }
}
