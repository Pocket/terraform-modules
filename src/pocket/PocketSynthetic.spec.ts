import { Testing } from 'cdktf';
import { PocketSyntheticCheck, PocketSyntheticProps } from './PocketSynthetic';

const config: PocketSyntheticProps = {
  config: {
    url: 'acme.getpocket.dev',
    path: '/healthz',
    port: '443',
    protocol: 'https:',
  },
  environment: 'test',
  artifactS3Location: 's3://not-a-real-thing',
};

test('renders a Pocket Synthetic check for acme.getpocket.dev', () => {
  const synthed = Testing.synthScope((stack) => {
    new PocketSyntheticCheck(stack, 'test-synthetic', config);
  });
  expect(synthed).toMatchSnapshot();
});

test('pass a completely custom test for synthetics', () => {
  delete config.config;
  config.check = `
  console.log('I am a synthetic-check');
`;
  const synthed = Testing.synthScope((stack) => {
    new PocketSyntheticCheck(stack, 'test-synthetic', config);
  });
  expect(synthed).toMatchSnapshot();
});
