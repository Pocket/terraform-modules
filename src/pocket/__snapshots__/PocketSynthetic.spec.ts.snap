// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pass a completely custom test for synthetics 1`] = `
"{
  \\"data\\": {
    \\"archive_file\\": {
      \\"test-synthetic_synthetic-zip-file_B5627663\\": {
        \\"output_path\\": \\"index-kw2pbiej.zip\\",
        \\"source_dir\\": \\"/home/bkochendorfer/work/pocket/terraform-modules/src/pocket/synthetics\\",
        \\"type\\": \\"zip\\"
      }
    },
    \\"aws_iam_policy_document\\": {
      \\"test-synthetic_assume-policy-document_4EBEEFDF\\": {
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"sts:AssumeRole\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"lambda.amazonaws.com\\",
                  \\"synthetics.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ]
          }
        ],
        \\"version\\": \\"2012-10-17\\"
      },
      \\"test-synthetic_execution-policy-document_E5D2936D\\": {
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"logs:CreateLogGroup\\",
              \\"logs:CreateLogStream\\",
              \\"logs:PutLogEvents\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"resources\\": [
              \\"arn:aws:logs:*:*:*\\"
            ]
          },
          {
            \\"actions\\": [
              \\"s3:ListAllMyBuckets\\",
              \\"cloudwatch:PutMetricData\\",
              \\"xray:PutTraceSegments\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"resources\\": [
              \\"*\\"
            ]
          },
          {
            \\"actions\\": [
              \\"s3:PutObject\\",
              \\"s3:GetBucketLocation\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"resources\\": [
              \\"arn:aws:s3:::not-a-real-thing\\",
              \\"arn:aws:s3:::not-a-real-thing/*\\"
            ]
          }
        ],
        \\"version\\": \\"2012-10-17\\"
      }
    }
  },
  \\"provider\\": {
    \\"archive\\": [
      {}
    ]
  },
  \\"resource\\": {
    \\"aws_cloudwatch_metric_alarm\\": {
      \\"test-synthetic_synthetics-failed-alarm_42A00967\\": {
        \\"alarm_actions\\": [],
        \\"alarm_description\\": \\"Synthetic Check is failing\\",
        \\"alarm_name\\": \\"test-synthetic-SyntheticsFailed-Alarm\\",
        \\"comparison_operator\\": \\"GreaterThanOrEqualToThreshold\\",
        \\"dimensions\\": {
          \\"CanaryName\\": \\"test-synthetic-synthetic\\"
        },
        \\"evaluation_periods\\": 1,
        \\"insufficient_data_actions\\": [],
        \\"metric_name\\": \\"Failures\\",
        \\"namespace\\": \\"CloudWatchSynthetics\\",
        \\"period\\": 60,
        \\"statistic\\": \\"SampleCount\\",
        \\"threshold\\": 1,
        \\"treat_missing_data\\": \\"notBreaching\\"
      }
    },
    \\"aws_iam_policy\\": {
      \\"test-synthetic_execution-policy_244BCA50\\": {
        \\"name\\": \\"test-synthetic-ExecutionRolePolicy\\",
        \\"policy\\": \\"\${data.aws_iam_policy_document.test-synthetic_execution-policy-document_E5D2936D.json}\\"
      }
    },
    \\"aws_iam_role\\": {
      \\"test-synthetic_synthetics-execution-role_3C0A4E8C\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test-synthetic_assume-policy-document_4EBEEFDF.json}\\",
        \\"name\\": \\"synthetics-role-for-test-synthetic\\"
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"test-synthetic_execution-role-policy-attachment_914E7B38\\": {
        \\"depends_on\\": [
          \\"aws_iam_role.test-synthetic_synthetics-execution-role_3C0A4E8C\\",
          \\"aws_iam_policy.test-synthetic_execution-policy_244BCA50\\"
        ],
        \\"policy_arn\\": \\"\${aws_iam_policy.test-synthetic_execution-policy_244BCA50.arn}\\",
        \\"role\\": \\"\${aws_iam_role.test-synthetic_synthetics-execution-role_3C0A4E8C.name}\\"
      }
    },
    \\"aws_synthetics_canary\\": {
      \\"test-synthetic_canary-test-synthetic_1B71309A\\": {
        \\"artifact_s3_location\\": \\"s3://not-a-real-thing\\",
        \\"execution_role_arn\\": \\"\${aws_iam_role.test-synthetic_synthetics-execution-role_3C0A4E8C.arn}\\",
        \\"handler\\": \\"index.handler\\",
        \\"name\\": \\"test-synthetic-syntheic\\",
        \\"runtime_version\\": \\"syn-nodejs-puppeteer-3.3\\",
        \\"schedule\\": {
          \\"expression\\": \\"rate(3 minutes)\\"
        },
        \\"start_canary\\": true,
        \\"zip_file\\": \\"\${data.archive_file.test-synthetic_synthetic-zip-file_B5627663.output_path}\\"
      }
    }
  },
  \\"terraform\\": {
    \\"required_providers\\": {
      \\"archive\\": {
        \\"source\\": \\"hashicorp/archive\\",
        \\"version\\": \\"~> 2.2\\"
      }
    }
  }
}"
`;

exports[`renders a Pocket Synthetic check for acme.getpocket.dev 1`] = `
"{
  \\"data\\": {
    \\"archive_file\\": {
      \\"test-synthetic_synthetic-zip-file_B5627663\\": {
        \\"output_path\\": \\"index-kw2pbi65.zip\\",
        \\"source_dir\\": \\"/home/bkochendorfer/work/pocket/terraform-modules/src/pocket/synthetics\\",
        \\"type\\": \\"zip\\"
      }
    },
    \\"aws_iam_policy_document\\": {
      \\"test-synthetic_assume-policy-document_4EBEEFDF\\": {
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"sts:AssumeRole\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"principals\\": [
              {
                \\"identifiers\\": [
                  \\"lambda.amazonaws.com\\",
                  \\"synthetics.amazonaws.com\\"
                ],
                \\"type\\": \\"Service\\"
              }
            ]
          }
        ],
        \\"version\\": \\"2012-10-17\\"
      },
      \\"test-synthetic_execution-policy-document_E5D2936D\\": {
        \\"statement\\": [
          {
            \\"actions\\": [
              \\"logs:CreateLogGroup\\",
              \\"logs:CreateLogStream\\",
              \\"logs:PutLogEvents\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"resources\\": [
              \\"arn:aws:logs:*:*:*\\"
            ]
          },
          {
            \\"actions\\": [
              \\"s3:ListAllMyBuckets\\",
              \\"cloudwatch:PutMetricData\\",
              \\"xray:PutTraceSegments\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"resources\\": [
              \\"*\\"
            ]
          },
          {
            \\"actions\\": [
              \\"s3:PutObject\\",
              \\"s3:GetBucketLocation\\"
            ],
            \\"effect\\": \\"Allow\\",
            \\"resources\\": [
              \\"arn:aws:s3:::not-a-real-thing\\",
              \\"arn:aws:s3:::not-a-real-thing/*\\"
            ]
          }
        ],
        \\"version\\": \\"2012-10-17\\"
      }
    }
  },
  \\"provider\\": {
    \\"archive\\": [
      {}
    ]
  },
  \\"resource\\": {
    \\"aws_cloudwatch_metric_alarm\\": {
      \\"test-synthetic_synthetics-failed-alarm_42A00967\\": {
        \\"alarm_actions\\": [],
        \\"alarm_description\\": \\"Synthetic Check is failing\\",
        \\"alarm_name\\": \\"test-synthetic-SyntheticsFailed-Alarm\\",
        \\"comparison_operator\\": \\"GreaterThanOrEqualToThreshold\\",
        \\"dimensions\\": {
          \\"CanaryName\\": \\"test-synthetic-synthetic\\"
        },
        \\"evaluation_periods\\": 1,
        \\"insufficient_data_actions\\": [],
        \\"metric_name\\": \\"Failures\\",
        \\"namespace\\": \\"CloudWatchSynthetics\\",
        \\"period\\": 60,
        \\"statistic\\": \\"SampleCount\\",
        \\"threshold\\": 1,
        \\"treat_missing_data\\": \\"notBreaching\\"
      }
    },
    \\"aws_iam_policy\\": {
      \\"test-synthetic_execution-policy_244BCA50\\": {
        \\"name\\": \\"test-synthetic-ExecutionRolePolicy\\",
        \\"policy\\": \\"\${data.aws_iam_policy_document.test-synthetic_execution-policy-document_E5D2936D.json}\\"
      }
    },
    \\"aws_iam_role\\": {
      \\"test-synthetic_synthetics-execution-role_3C0A4E8C\\": {
        \\"assume_role_policy\\": \\"\${data.aws_iam_policy_document.test-synthetic_assume-policy-document_4EBEEFDF.json}\\",
        \\"name\\": \\"synthetics-role-for-test-synthetic\\"
      }
    },
    \\"aws_iam_role_policy_attachment\\": {
      \\"test-synthetic_execution-role-policy-attachment_914E7B38\\": {
        \\"depends_on\\": [
          \\"aws_iam_role.test-synthetic_synthetics-execution-role_3C0A4E8C\\",
          \\"aws_iam_policy.test-synthetic_execution-policy_244BCA50\\"
        ],
        \\"policy_arn\\": \\"\${aws_iam_policy.test-synthetic_execution-policy_244BCA50.arn}\\",
        \\"role\\": \\"\${aws_iam_role.test-synthetic_synthetics-execution-role_3C0A4E8C.name}\\"
      }
    },
    \\"aws_synthetics_canary\\": {
      \\"test-synthetic_canary-test-synthetic_1B71309A\\": {
        \\"artifact_s3_location\\": \\"s3://not-a-real-thing\\",
        \\"execution_role_arn\\": \\"\${aws_iam_role.test-synthetic_synthetics-execution-role_3C0A4E8C.arn}\\",
        \\"handler\\": \\"index.handler\\",
        \\"name\\": \\"test-synthetic-syntheic\\",
        \\"runtime_version\\": \\"syn-nodejs-puppeteer-3.3\\",
        \\"schedule\\": {
          \\"expression\\": \\"rate(3 minutes)\\"
        },
        \\"start_canary\\": true,
        \\"zip_file\\": \\"\${data.archive_file.test-synthetic_synthetic-zip-file_B5627663.output_path}\\"
      }
    }
  },
  \\"terraform\\": {
    \\"required_providers\\": {
      \\"archive\\": {
        \\"source\\": \\"hashicorp/archive\\",
        \\"version\\": \\"~> 2.2\\"
      }
    }
  }
}"
`;
